### PR TITLE
[HIPPEROS] Configuration and minimal syscall implementation

### PR DESCRIPTION
**Pull request by odesenfans**
Added support for the `*-*-hipperos*` triplets in newlib. This configuration
builds `newlib` for the HIPPEROS RTOS.

Implemented minimal versions of all required (reentrant) syscalls.
The implementations of `_getpid_r`, `_exit` and `_printf_r` are
functional.

### CODE REVIEW

File: newlib/configure.host [hunk 1]
```diff
@@ -385,6 +385,9 @@ case "${host}" in
 	signal_dir=
 	sys_dir=netware
 	;;
+  *-*-hipperos*)
```
Comment (VladimirSvoboda) [inline newlib/configure.host]: > THIS TABLE IS ALPHA SORTED. KEEP IT THAT WAY.

File: newlib/libc/sys/configure.in [hunk 2]
```diff
@@ -36,6 +36,7 @@ if test -n "${sys_dir}"; then
 	or1k) AC_CONFIG_SUBDIRS(or1k) ;;
 	phoenix) AC_CONFIG_SUBDIRS(phoenix) ;;
 	rdos) AC_CONFIG_SUBDIRS(rdos) ;;
+    hipperos) AC_CONFIG_SUBDIRS(hipperos) ;;
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/configure.in]: Should start with a TAB and alpha sorted.

File: newlib/libc/sys/hipperos/crt0.c [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#include <fcntl.h>
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/crt0.c]: Why?

File: newlib/libc/sys/hipperos/syscalls.c [hunk 4]
```diff
@@ -0,0 +1,130 @@
+#include <reent.h>
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/syscalls.c:1]: I guess that we will have to add license stuff.

File: newlib/libc/sys/hipperos/syscalls.c [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+{
+    return -1;
+}
+
+_ssize_t _read_r(struct _reent* r, int file, void* ptr, size_t len)
+{
+    return 0;
+}
+
+void* _sbrk_r(struct _reent* r, ptrdiff_t incr)
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/syscalls.c:95]: I think we should try to not rely on this as this is deprecated.

File: newlib/libc/sys/hipperos/syscalls.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    return -1;
+}
+
+int _fstat_r(struct _reent* r, int file, struct stat* st)
+{
+    st->st_mode = S_IFCHR;
+    return 0;
+}
+
+int _getpid_r(struct _reent* r)
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/syscalls.c:57]: Move the implemented functions to their own files and keep the "unimplemented.c" file as we do usually.

File: newlib/libc/sys/hipperos/syscalls.c [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    static char* heap_end;
+    char* prev_heap_end;
+
+    if (heap_end == 0) {
+        heap_end = &_app_bss_end;
+    }
+    prev_heap_end = heap_end;
+    if (heap_end + incr > stack_ptr) {
+        _write_r(r, 1, "Heap and stack collision\n", 25);
+        abort();
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/syscalls.c:108]: What does abort do?

**Review by VladimirSvoboda — CHANGES_REQUESTED** [review]
Some small changes. Try to use the same version of `auto*` as them.

File: newlib/configure.host (hunk 1 above)
Comment (odesenfans) [reply newlib/configure.host]: Oops thought I fixed that. My bad.

File: newlib/libc/sys/hipperos/syscalls.c (hunk 4 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:1]: Yes, I don't know what yet.

File: newlib/libc/sys/hipperos/syscalls.c (hunk 5 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:95]: Yup. It's required by their implementation of `malloc` so the clean fix is to provide our implementation of `malloc`.

File: newlib/libc/sys/hipperos/syscalls.c (hunk 7 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:108]: Calls `exit(EXIT_FAILURE)`.

File: newlib/configure.host (hunk 1 above)
Comment (odesenfans) [reply newlib/configure.host]: Fixed in 2b9590ca6fbd25e8e78aad87627eae115f863b6d.

File: newlib/libc/sys/configure.in (hunk 2 above)
Comment (odesenfans) [reply newlib/libc/sys/configure.in]: 2b9590ca6fbd25e8e78aad87627eae115f863b6d

File: newlib/libc/sys/hipperos/crt0.c (hunk 3 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/crt0.c]: Useless. Fixed in 2b9590ca6fbd25e8e78aad87627eae115f863b6d

File: newlib/libc/sys/hipperos/syscalls.c (hunk 6 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:57]: Will do that in future revisions.

File: newlib/libc/sys/hipperos/syscalls.c (hunk 7 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:108]: The actual implementation is:

```c
_VOID
_DEFUN_VOID (abort)
{
#ifdef ABORT_MESSAGE
  write (2, "Abort called\n", sizeof ("Abort called\n")-1);
#endif

  while (1)
    {
      raise (SIGABRT);
      _exit (1);
    }
}
```

File: newlib/libc/sys/hipperos/configure [hunk 8]
```diff
@@ -0,0 +1,4716 @@
+#! /bin/sh
+# Guess values for system-dependent variables and create Makefiles.
+# Generated by GNU Autoconf 2.64 for newlib 2.5.0.
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/configure]: Use `autoconf` 2.68 here too (this is what rtems and phoenix do).

File: newlib/libc/sys/hipperos/syscalls.c [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                 uint16_t messageType,
+                 const char* contentPtr,
+                 uint32_t contentLength);
+
+
+#undef errno
+extern int errno;
+
+char* __env[1] = {0};
+char** environ = __env;
```
Comment (paul-rodriguez) [inline newlib/libc/sys/hipperos/syscalls.c:23]: What's the point of these two lines?

**Review by paul-rodriguez — APPROVED** [review]
I like turtles

File: newlib/libc/sys/hipperos/syscalls.c (hunk 9 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/syscalls.c:23]: Those two are the representation of environment variable and must be declared here.

**Comment by apaolillo** [comment]
How often will we rebase our work on the bare newlib-cygwin repo?

**Review by apaolillo — APPROVED** [review]
Looks ok to me

**Comment by odesenfans** [comment]
@apaolillo I'd say for each major release, about once a year.

**Review by VladimirSvoboda — APPROVED** [review]

**Pull request merged**